arangod/MMFiles/MMFilesCollection.cpp

```diff
@@ -3053,6 +3056,149 @@ void MMFilesCollection::removeLocalDocumentId(LocalDocumentId const& documentId,
   }
 }
 
+namespace {
+Result countDocumentsIterator(MMFilesMarker const* marker,
+                              void* counter, MMFilesDatafile*) {
+  Result res;
+  TRI_ASSERT(nullptr != counter);
+  if (marker->getType() == TRI_DF_MARKER_VPACK_DOCUMENT) {
+    (*static_cast<int*>(counter))++;
+  }
+  return res;
+}
+
+Result persistLocalDocumentIdIterator(
+    MMFilesMarker const* marker, void* data, MMFilesDatafile* inputFile) {
+  Result res;
+  auto outputFile = static_cast<MMFilesDatafile*>(data);
+  switch (marker->getType()) {
+    case TRI_DF_MARKER_VPACK_DOCUMENT: {
+      auto transactionId = MMFilesDatafileHelper::TransactionId(marker);
+
+      VPackSlice const slice(
+          reinterpret_cast<char const*>(marker) +
+          MMFilesDatafileHelper::VPackOffset(TRI_DF_MARKER_VPACK_DOCUMENT));
+      uint8_t const* vpack = slice.begin();
+
+      LocalDocumentId localDocumentId;
+      if (marker->getSize() ==
+          MMFilesDatafileHelper::VPackOffset(TRI_DF_MARKER_VPACK_DOCUMENT)
+          + slice.byteSize() + sizeof(LocalDocumentId::BaseType)) {
+        // we do have a LocalDocumentId stored at the end of the marker
+        uint8_t const* ptr = vpack + slice.byteSize();
+        localDocumentId = LocalDocumentId(
+            encoding::readNumber<LocalDocumentId::BaseType>(
+                ptr, sizeof(LocalDocumentId::BaseType)));
+      } else {
+        localDocumentId = LocalDocumentId::create();
+      }
+
+      MMFilesCrudMarker updatedMarker(TRI_DF_MARKER_VPACK_DOCUMENT,
+                                      transactionId, localDocumentId, slice);
+
+      std::unique_ptr<char> buffer(new char[updatedMarker.size()]);
```

Comment:
This should be a `unique_ptr<char[]>` to ensure that the correct operator delete is used. Since you are now using C++14 you could simply do `auto buffer = std::make_unique<char[]>(updatedMarker.size())`.

Reply:
@mpoeter : agree on `char[]`, but disagree on C++14 :smile:
 
We have moved to C++14, but only for our devel branch. This is the 3.4 branch, where we will still use C++11.

Reply:
@mpoeter : thanks for reporting this, would have been some not-easy to find issue because it may have seemed to work by chance but with unforeseeable runtime effects.

Reply:
@jsteemann Too bad! :wink: I saw the PR to move to C++14 but did not realize that this was only on devel.